File: backend/agent/bash.go
package agent

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"strings"
	"sync"
	"time"

	"github.com/enough/enough/backend/opencode"
	"github.com/enough/enough/backend/shell"
)

const (
	maxBashOutput        = 32_000
	exitStdioGrace       = 100 * time.Millisecond // Flame: don't hang on inherited stdio after shell exit
	bashUpdateThrottle   = 100 * time.Millisecond
)

func bashTool() opencode.Tool {
	return opencode.Tool{
		Type: "function",
		Function: opencode.ToolFunction{
			Name:        "bash",
			Description: "Run a shell command in the project workspace. Do NOT run mpv, sixel, blessed, or full-screen TUI apps — they break the Enough terminal. Use curl, tests, and plain-text commands only.",
			Parameters: json.RawMessage(`{
				"type": "object",
				"properties": {
					"command": {"type": "string"}
				},
				"required": ["command"]
			}`),
		},
	}
}

func (a *Agent) toolBash(ctx context.Context, id, argsJSON string) toolResult {
	var args struct {
		Command string `json:"command"`
	}
	if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
		return toolResult{output: err.Error(), isErr: true}
	}

	if blocked := bashCommandBlocked(args.Command); blocked != "" {
		return toolResult{output: blocked, isErr: true}
	}

	// CommandContext + the platform Cancel hook means an aborted context (ESC)
	// actually kills the running process (and its group on unix), instead of
	// the command running to completion after cancellation.
	cmd, err := shell.CommandContext(ctx, args.Command, true)
	if err != nil {
		return toolResult{output: err.Error(), isErr: true}
	}
	cmd.Dir = shell.ResolveSafeCwd(a.workDir)
	configureProcGroup(cmd)

	delta := newBashDeltaEmitter(func(chunk string) {
		clean, _ := SanitizeBashOutput(chunk)
		if clean != "" {
			a.toolDelta(id, clean)
		}
	})
	defer delta.flush()

	sw := &bashStreamWriter{max: maxBashOutput, onChunk: delta.add}

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return toolResult{output: err.Error(), isErr: true}
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return toolResult{output: err.Error(), isErr: true}
	}

	if err := cmd.Start(); err != nil {
		return toolResult{output: err.Error(), isErr: true}
	}
	a.registerBashCmd(cmd)
	defer a.unregisterBashCmd(cmd)

	var wg sync.WaitGroup
	wg.Add(2)
	copyOut := func(r io.ReadCloser) {
		defer wg.Done()
		_, _ = io.Copy(sw, r)
	}
	go copyOut(stdout)
	go copyOut(stderr)

	started := time.Now()
	copyDone := make(chan struct{})
	go func() {
		wg.Wait()
		close(copyDone)
	}()

	exitDone := make(chan error, 1)
	go func() {
		exitDone <- cmd.Wait()
	}()

	closeReadersAfterGrace := func() {
		select {
		case <-copyDone:
		case <-time.After(exitStdioGrace):
			_ = stdout.Close()
			_ = stderr.Close()
			<-copyDone
		}
	}

	var waitErr error
	select {
	case err := <-exitDone:
		waitErr = err
		closeReadersAfterGrace()
	case <-copyDone:
		waitErr = <-exitDone
	case <-ctx.Done():
		_ = killProcessGroup(cmd)
		closeReadersAfterGrace()
		waitErr = <-exitDone
	}

	sw.Finalize()
	delta.flush()
	duration := time.Since(started)
	text, _ := SanitizeBashOutput(sw.String())

	// Interrupted: report whatever was captured plus a clear marker rather than
	// a raw "signal: killed" error. No evidence — an interrupted run proves nothing.
	if ctx.Err() != nil {
		if text != "" && !strings.HasSuffix(text, "\n") {
			text += "\n"
		}
		return toolResult{output: text + "[interrupted]", isErr: true}
	}

	exitCode := 0
	if waitErr != nil {
		exitCode = -1
		var ee *exec.ExitError
		if errors.As(waitErr, &ee) {
			exitCode = ee.ExitCode()
		}
	}
	a.recordCommandRun(args.Command, exitCode, text, duration)

	if waitErr != nil {
		return toolResult{output: fmt.Sprintf("%v\n%s", waitErr, text), isErr: true}
	}
	return toolResult{output: text}
}

func (a *Agent) registerBashCmd(cmd *exec.Cmd) {
	a.activeBashMu.Lock()
	a.activeBashCmd = cmd
	a.activeBashMu.Unlock()
}

func (a *Agent) unregisterBashCmd(cmd *exec.Cmd) {
	a.activeBashMu.Lock()
	if a.activeBashCmd == cmd {
		a.activeBashCmd = nil
	}
	a.activeBashMu.Unlock()
}

func (a *Agent) killActiveBash() {
	a.activeBashMu.Lock()
	cmd := a.activeBashCmd
	a.activeBashMu.Unlock()
	if cmd != nil {
		_ = killProcessGroup(cmd)
	}
}

// bashStreamWriter accumulates command output up to a cap while streaming each
// appended chunk to onChunk. The streamed chunks concatenate exactly to the
// final String(), so the live view and the persisted result stay consistent.
type bashStreamWriter struct {
	mu        sync.Mutex
	buf       strings.Builder
	max       int
	total     int
	truncated bool
	onChunk   func(string)
}

const truncMarker = "\n... truncated ..."

func (w *bashStreamWriter) Write(p []byte) (int, error) {
	w.mu.Lock()
	w.total += len(p)
	var emit string
	if !w.truncated {
		room := w.max - w.buf.Len()
		switch {
		case room <= 0:
			w.truncated = true
			w.buf.WriteString(truncMarker)
			emit = truncMarker
		case len(p) <= room:
			clean, _ := SanitizeBashOutput(string(p))
			w.buf.WriteString(clean)
			emit = clean
		default:
			clean, _ := SanitizeBashOutput(string(p[:room]))
			w.buf.WriteString(clean)
			w.buf.WriteString(truncMarker)
			w.truncated = true
			emit = clean + truncMarker
		}
	}
	w.mu.Unlock()

	if emit != "" && w.onChunk != nil {
		w.onChunk(emit)
	}
	return len(p), nil
}

func (w *bashStreamWriter) Finalize() {
	w.mu.Lock()
	var emit string
	if w.total > w.max && !w.truncated {
		w.truncated = true
		w.buf.WriteString(truncMarker)
		emit = truncMarker
	}
	w.mu.Unlock()
	if emit != "" && w.onChunk != nil {
		w.onChunk(emit)
	}
}

func (w *bashStreamWriter) String() string {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.buf.String()
}

// bashDeltaEmitter throttles live bash output updates (Flame uses 100ms) so the
// TUI event channel is not flooded and copy goroutines cannot block on emit.
type bashDeltaEmitter struct {
	mu      sync.Mutex
	emit    func(string)
	pending strings.Builder
	lastAt  time.Time
	timer   *time.Timer
}

func newBashDeltaEmitter(emit func(string)) *bashDeltaEmitter {
	return &bashDeltaEmitter{emit: emit}
}

func (e *bashDeltaEmitter) add(chunk string) {
	if chunk == "" || e.emit == nil {
		return
	}
	e.mu.Lock()
	defer e.mu.Unlock()
	e.pending.WriteString(chunk)
	delay := bashUpdateThrottle - time.Since(e.lastAt)
	if delay <= 0 {
		e.flushLocked()
		return
	}
	if e.timer == nil {
		e.timer = time.AfterFunc(delay, func() {
			e.mu.Lock()
			defer e.mu.Unlock()
			e.timer = nil
			e.flushLocked()
		})
	}
}

func (e *bashDeltaEmitter) flush() {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.timer != nil {
		e.timer.Stop()
		e.timer = nil
	}
	e.flushLocked()
}

func (e *bashDeltaEmitter) flushLocked() {
	if e.pending.Len() == 0 {
		return
	}
	e.emit(e.pending.String())
	e.pending.Reset()
	e.lastAt = time.Now()
}

File: backend/agent/bash.ts
// PORT: backend/agent/bash.go

import { Effect } from "effect";
import { type tool } from "../opencode/types";
import { Agent, type toolResult } from "./agent";
import { command_context } from "../shell/command";
import { resolve_safe_cwd } from "../shell/cwd";
import {
  trackDetachedChildPid,
  untrackDetachedChildPid,
  killTrackedDetachedChildren,
  installDetachedShutdownHook,
} from "../shell/process_tree";
import { bashCommandBlocked, SanitizeBashOutput } from "./bash_sanitize";
import { configureProcGroup as linuxConfigure, killProcessGroup as linuxKill } from "./bash_linux";
import { configureProcGroup as unixConfigure, killProcessGroup as unixKill } from "./bash_unix";
import { configureProcGroup as windowsConfigure, killProcessGroup as windowsKill } from "./bash_windows";
import { configureProcGroup as otherConfigure, killProcessGroup as otherKill } from "./bash_other";
import process from "node:process";

// Reap detached children when the backend exits naturally.
installDetachedShutdownHook();

const maxBashOutput = 32000;
const exitStdioGrace = 100; // ms
const bashUpdateThrottle = 100; // ms
const truncMarker = "\n... truncated ...";

function configureProcGroup(cmd: any): void {
  if (process.platform === "linux") {
    linuxConfigure(cmd);
  } else if (process.platform === "win32") {
    windowsConfigure(cmd);
  } else if (
    process.platform === "darwin" ||
    process.platform === "freebsd" ||
    process.platform === "openbsd" ||
    process.platform === "netbsd"
  ) {
    unixConfigure(cmd);
  } else {
    otherConfigure(cmd);
  }
}

function killProcessGroup(cmd: any): Error | null {
  if (process.platform === "linux") {
    return linuxKill(cmd);
  } else if (process.platform === "win32") {
    return windowsKill(cmd);
  } else if (
    process.platform === "darwin" ||
    process.platform === "freebsd" ||
    process.platform === "openbsd" ||
    process.platform === "netbsd"
  ) {
    return unixKill(cmd);
  } else {
    return otherKill(cmd);
  }
}

export function bashTool(): tool {
  const schema = {
    type: "object",
    properties: {
      command: { type: "string" },
    },
    required: ["command"],
  };
  return {
    type: "function",
    function: {
      name: "bash",
      description: "Run a shell command in the project workspace. Do NOT run mpv, sixel, blessed, or full-screen TUI apps — they break the Hollow terminal. Use curl, tests, and plain-text commands only.",
      parameters: new TextEncoder().encode(JSON.stringify(schema)),
    },
  };
}

Agent.prototype.toolBash = function (
  this: Agent,
  ctx: AbortSignal,
  id: string,
  argsJSON: string
): Effect.Effect<toolResult, Error> {
  let args: { command: string };
  try {
    args = JSON.parse(argsJSON);
  } catch (err) {
    return Effect.succeed({ output: err instanceof Error ? err.message : String(err), isErr: true });
  }

  const blocked = bashCommandBlocked(args.command);
  if (blocked !== "") {
    return Effect.succeed({ output: blocked, isErr: true });
  }

  const safeDir = resolve_safe_cwd(this.workDir);
  const commandEff = command_context(ctx, args.command, true, safeDir);

  return commandEff.pipe(
    Effect.flatMap((child) => {
      configureProcGroup(child);
      if (child.pid) trackDetachedChildPid(child.pid);

      const delta = new BashDeltaEmitter((chunk) => {
        const [clean] = SanitizeBashOutput(chunk);
        if (clean !== "") {
          try {
            if (this.emit) {
              this.emit({ kind: "tool_delta", data: { id, chunk: clean } });
            }
          } catch {}
        }
      });

      const sw = new BashStreamWriter(maxBashOutput, (chunk) => delta.add(chunk));

      child.stdout?.on("data", (chunk: Buffer) => {
        sw.write(chunk.toString("utf8"));
      });
      child.stderr?.on("data", (chunk: Buffer) => {
        sw.write(chunk.toString("utf8"));
      });

      const started = Date.now();
      this.registerBashCmd(child);

      // waitForChildProcess: resolve once the process has exited AND its stdio
      // has ended, with a short grace backstop so inherited stdio handles held
      // by detached descendants can't hang us forever. Mirrors Flame's
      // utils/child-process.ts waitForChildProcess.
      return Effect.async<toolResult, Error>((resume) => {
        let settled = false;
        let exited = false;
        let exitCode: number | null = null;
        let stdoutEnded = child.stdout === null || child.stdout === undefined;
        let stderrEnded = child.stderr === null || child.stderr === undefined;
        let postExitTimer: NodeJS.Timeout | undefined;

        const cleanup = () => {
          if (postExitTimer) {
            clearTimeout(postExitTimer);
            postExitTimer = undefined;
          }
          child.removeListener("error", onError);
          child.removeListener("exit", onExit);
          child.removeListener("close", onClose);
          child.stdout?.removeListener("end", onStdoutEnd);
          child.stderr?.removeListener("end", onStderrEnd);
          ctx.removeEventListener("abort", onAbort);
          if (child.pid) untrackDetachedChildPid(child.pid);
          this.unregisterBashCmd(child);
        };

        const finalize = (code: number | null) => {
          if (settled) return;
          settled = true;
          cleanup();
          sw.finalize();
          delta.flush();
          child.stdout?.destroy();
          child.stderr?.destroy();

          const duration = Date.now() - started;
          const [text] = SanitizeBashOutput(sw.toString());

          if (ctx.aborted) {
            const out = text === "" ? "" : text.endsWith("\n") ? text : text + "\n";
            resume(Effect.succeed({ output: out + "Command aborted", isErr: true }));
            return;
          }

          this.recordCommandRun(args.command, code ?? -1, text, duration);

          if (code !== null && code !== 0) {
            const out = text === "" ? "" : text.endsWith("\n") ? text : text + "\n";
            resume(Effect.succeed({ output: out + `Command exited with code ${code}`, isErr: true }));
            return;
          }
          resume(Effect.succeed({ output: text }));
        };

        const maybeFinalizeAfterExit = () => {
          if (!exited || settled) return;
          if (stdoutEnded && stderrEnded) finalize(exitCode);
        };

        const onStdoutEnd = () => {
          stdoutEnded = true;
          maybeFinalizeAfterExit();
        };
        const onStderrEnd = () => {
          stderrEnded = true;
          maybeFinalizeAfterExit();
        };
        const onExit = (code: number | null) => {
          exited = true;
          exitCode = code;
          maybeFinalizeAfterExit();
          if (!settled) {
            postExitTimer = setTimeout(() => finalize(code), exitStdioGrace);
          }
        };
        const onClose = (code: number | null) => finalize(code);
        const onError = (err: Error) => {
          if (settled) return;
          settled = true;
          cleanup();
          resume(Effect.succeed({ output: err.message, isErr: true }));
        };
        const onAbort = () => {
          killProcessGroup(child);
          finalize(exitCode);
        };

        child.stdout?.once("end", onStdoutEnd);
        child.stderr?.once("end", onStderrEnd);
        child.once("error", onError);
        child.once("exit", onExit);
        child.once("close", onClose);
        if (ctx.aborted) {
          onAbort();
        } else {
          ctx.addEventListener("abort", onAbort);
        }
      });
    }),
    Effect.catchAll((err: any) =>
      Effect.succeed({ output: err instanceof Error ? err.message : String(err), isErr: true })
    )
  );
};

Agent.prototype.registerBashCmd = function (this: Agent, cmd: any) {
  this.activeBashCmd = cmd;
};

Agent.prototype.unregisterBashCmd = function (this: Agent, cmd: any) {
  if (this.activeBashCmd === cmd) {
    this.activeBashCmd = null;
  }
};

Agent.prototype.killActiveBash = function (this: Agent) {
  const cmd = this.activeBashCmd;
  if (cmd !== null) {
    killProcessGroup(cmd);
  }
  // Reap any other detached children still tracked (e.g. descendants that
  // outlived their parent). The active cmd is also in the set; re-killing a
  // dead pid is a no-op.
  killTrackedDetachedChildren();
};

export class BashStreamWriter {
  private buf: string[] = [];
  private max: number;
  private total = 0;
  private truncated = false;
  private onChunk?: (chunk: string) => void;
  private currentLen = 0;

  constructor(max: number, onChunk?: (chunk: string) => void) {
    this.max = max;
    this.onChunk = onChunk;
  }

  write(chunk: string): void {
    this.total += chunk.length;
    let emit = "";
    if (!this.truncated) {
      const room = this.max - this.currentLen;
      if (room <= 0) {
        this.truncated = true;
        this.buf.push(truncMarker);
        this.currentLen += truncMarker.length;
        emit = truncMarker;
      } else if (chunk.length <= room) {
        const [clean] = SanitizeBashOutput(chunk);
        this.buf.push(clean);
        this.currentLen += clean.length;
        emit = clean;
      } else {
        const [clean] = SanitizeBashOutput(chunk.slice(0, room));
        this.buf.push(clean);
        this.buf.push(truncMarker);
        this.currentLen += clean.length + truncMarker.length;
        this.truncated = true;
        emit = clean + truncMarker;
      }
    }
    if (emit !== "" && this.onChunk) {
      this.onChunk(emit);
    }
  }

  finalize(): void {
    let emit = "";
    if (this.total > this.max && !this.truncated) {
      this.truncated = true;
      this.buf.push(truncMarker);
      this.currentLen += truncMarker.length;
      emit = truncMarker;
    }
    if (emit !== "" && this.onChunk) {
      this.onChunk(emit);
    }
  }

  toString(): string {
    return this.buf.join("");
  }
}

export class BashDeltaEmitter {
  private emit: (chunk: string) => void;
  private pending: string[] = [];
  private lastAt = 0;
  private timer: NodeJS.Timeout | null = null;

  constructor(emit: (chunk: string) => void) {
    this.emit = emit;
  }

  add(chunk: string): void {
    if (chunk === "") {
      return;
    }
    this.pending.push(chunk);
    const now = Date.now();
    const delay = bashUpdateThrottle - (now - this.lastAt);
    if (delay <= 0) {
      this.flushLocked();
      return;
    }
    if (this.timer === null) {
      this.timer = setTimeout(() => {
        this.timer = null;
        this.flushLocked();
      }, delay);
    }
  }

  flush(): void {
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.flushLocked();
  }

  private flushLocked(): void {
    if (this.pending.length === 0) {
      return;
    }
    this.emit(this.pending.join(""));
    this.pending = [];
    this.lastAt = Date.now();
  }
}

/*
PORT STATUS
source path: backend/agent/bash.go
source lines: 304
draft lines: 279
confidence: high
status: phase_b_compile
*/
